Source/WebCore/platform/graphics/cg/ImageBufferIOSurfaceBackend.cpp

```diff
@@ -274,23 +275,54 @@ void ImageBufferIOSurfaceBackend::prepareForExternalRead()
 {
     // Ensure that there are no pending draws to this surface. This is ensured by flushing the context
     // through which the draws may have come.
-    flushContext();
+    flushContextIfNeeded();
 }
 
 void ImageBufferIOSurfaceBackend::prepareForExternalWrite()
 {
+    bool needFlush = false;
     // Ensure that there are no future draws from the surface that would use the surface context image cache.
     if (m_mayHaveOutstandingBackingStoreReferences) {
         invalidateCachedNativeImage();
         m_mayHaveOutstandingBackingStoreReferences = false;
+        needFlush = true;
     }
 
     // Ensure that there are no pending draws to this surface. This is ensured by flushing the context
     // through which the draws may have come.
     // Ensure that there are no pending draws from this surface. This is ensured by drawing the invalidation marker before
     // flushing the the context. The invalidation marker forces the draws from this surface to complete before
     // the invalidation marker completes.
-    flushContext();
+    if (flushContextIfNeeded())
```

Comment:
It's pretty confusing that this function calls both `flushContextIfNeeded()` and `CGContextFlush()`.

Reply:
Renamed `flushContextIfNeeded` to `flushContextIfHasDrawn`